README.rst
========
Squarish
========


.. image:: https://img.shields.io/pypi/v/squarish.svg
        :target: https://pypi.python.org/pypi/squarish

.. image:: https://img.shields.io/travis/adammichaelwood/squarish.svg
        :target: https://travis-ci.org/adammichaelwood/squarish

.. image:: https://readthedocs.org/projects/squarish/badge/?version=latest
        :target: https://squarish.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status




Squarish is a small utility library for finding and working with integers whose factors neapprovximate useful ratios.


* Free software: MIT license
* Documentation: https://squarish.readthedocs.io.


Features
--------

* TODO

Credits
-------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.0.0 (2018-08-27)
------------------

* First release on PyPI.


tests/test_squarish.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `squarish` package."""

import pytest

from click.testing import CliRunner

from squarish import squarish
from squarish import cli


@pytest.fixture
def response():
    """Sample pytest fixture.

    See more at: http://doc.pytest.org/en/latest/fixture.html
    """
    # import requests
    # return requests.get('https://github.com/audreyr/cookiecutter-pypackage')


def test_content(response):
    """Sample pytest test function with the pytest fixture as an argument."""
    # from bs4 import BeautifulSoup
    # assert 'GitHub' in BeautifulSoup(response.content).title.string


def test_command_line_interface():
    """Test the CLI."""
    runner = CliRunner()
    result = runner.invoke(cli.main)
    assert result.exit_code == 0
    assert 'squarish.cli.main' in result.output
    help_result = runner.invoke(cli.main, ['--help'])
    assert help_result.exit_code == 0
    assert '--help  Show this message and exit.' in help_result.output


squarish/arbitrary_ratios.py
import math
import functools
import copy
import random

def fittest_total(m,n,r,return_greatest=False):
    """Returns the integer in the range `(m, m + n)` (inclusive),
    which has a pair of factors whose ratio is nearest r.`
    
    In the case of several equally fit possibilities,
    the lowest is returned by default.

    Parameters
    ----------

    m : int
        The low end of the range.

    n : int
        The maximum value that can be added to `m`,
        to produce a fit number.

    r : number or tuple or fraction
        A ratio to fit.

    return_greatest : bool, optional
        Whether to return the lowest (False, default) or highest (True) 
        in the case of multiple equally-fit numbers within range.

    Examples
    --------

    >>> fittest_total(8,4,2)
    8

    >>> fittest_total(8,4,(3,4))
    12

    """

    return fittest_of(range(m,m+n+1), r, reverse_on_key=return_greatest)
        

def fittest_of(c, r, sort_key=None, reverse_on_key=False):
    """Returns the :term:`fittest` member of `c`, a collection of numbers.

    If there are multiple equally fit numbers,
    `sort_key` is used to pick a single one.

    Parameters
    ----------

    c : iterable of int
        A collection of integers.

    r : number or tuple or fraction
        A ratio to fit.

    sort_key : callable, optional
        Callable used to sort the results 
        in the case of multiple equally-fit numbers.

        By default, the function returns the lowest candidate.

    reverse_on_key : bool, optional
        Whether to reverse the sort order
        in the case of multiple candidates.
    
        When ``False`` is used with the default ``sort_key``,
        the highest fit number will be returned.

    Examples
    --------

    >>> fittest_of([1,2,3,4,5,6,7,8,9,10,11,12], 2)
    2

    >>> fittest_of([1,2,3,4,5,6,7,8,9,10,11,12], 2, reverse_on_key=True])
    8

    >>> fittest_of([8,10,15,17,24,26,35,37,48,50,63,65], (3,4))
    48

    """

    raise NotImplementedError


@functools.lru_cache()
def factors_of(x):
    """Returns a set of tuples of factors of x.

    Examples
    --------

    >>> factors_of(12)
    {(1, 12), (2, 6), (3, 4)}

    >>> factors_of(20)
    {(1, 20), (2, 10), (4,5)}
    """

    factors = set()
    sqrt_x = math.sqrt(x)
    if sqrt_x == int(sqrt_x):
        factors.add((sqrt_x, sqrt_x))

    for d in range(math.floor(sqrt_x)): # d for divisor
        if x % d == 0:
            factors.add((d,int(x/d)))

    return factors
    



    
@functools.lru_cache()
def fittest_factors(x, r):
    """Returns the pair of factors of x,
       whose ratio is closest to r.

    Parameters
    ----------

    x : int
    r : number or fraction or tuple

    Examples
    --------

    >>> fittest_factors(20, (6,4))
    (5, 4)

    >>> fittest_factors(36, 1)
    (6, 6)

    >>> squarish_factors(53, 0.9)
    (53, 1)
    """

    raise NotImplementedError

def is_squarish(x, threshold=0.75):
    """Determines whether x is sufficiently squarish.

    Parameters
    ----------

    x : int

    threshold : float, optional
        The lower bound of acceptable squareness.

    Examples
    --------

    >>> is_squarish(67)
    False

    >>> is_squarish(12)
    True
    """

    return squareness(x) > threshold


def squareness(x):
    """Returns a number between ``(0,1)`` 
    representing the squareness of ``x``.

    All perfect squares have a squareness of ``1``.
    

    """

    
    if x == 0:
        return 1
    sr = math.sqrt(x)
    m,n = squarish_factors(x)
    
    return (sr/m + n/sr)/2

    
    


###

def dict_list_append(d, index_of_list, item):
    """Adds an item to a list in a dictionary,
    and creates the list if needed.

    Returns nothing.
    """
    try:
        d[index_of_list].append(item)
    except IndexError:
        d[index_of_list] = [item]


def r_size(x):
    """Returns the difference between the largest and smallest value in a collection."""
    return max(x) - min(x)

def is_square(x):
    """Returns True if x is a perfect square. Otherwise False."""
    return int(math.sqrt(x)) == math.sqrt(x)


def is_prime(x):
    """Returns True if x is prime. Otherwise False."""

    return min(squarish_factors(x)) == 1



squarish/squarish.py
# -*- coding: utf-8 -*-

"""Main module."""

import math
import functools

def squarest_total(m,n,return_greatest=False):
    """Returns the :term:`squarest` integer
    in the range `(m, m + n)` (inclusive).
    
    In the case of several perfect squares,
    the lowest is returned by default.

    Parameters
    ----------

    m : int
        The low end of the range.

    n : int
        The maximum value that can be added to `m`,
        to produce a squarish number.

    return_greatest : bool, optional
        Whether to return the lowest (False, default) or highest (True) 
        in the case of multiple perfect squares within range.

    Examples
    --------

    >>> squarest_total(10,7)
    16

    >>> squarest_total(3,8)
    4

    >>> squarest_total(3,8,return_greatest=True)
    9

    >>> squarest_total(9,27)
    9

    >>> squarest_total(9,27,return_greatest=True)
    36

    >>> squarest_total(37, 10)
    42

    """

    return squarest_of(range(m,m+n+1), reverse_on_key=return_greatest)
        

def squarest_of(c, sort_key=None, reverse_on_key=False):
    """Returns the :term:`squarest` member of `c`, a collection of numbers.

    If there are multiple perfect squares, 
    `sort_key` is used to pick a single one.

    Parameters
    ----------

    c : iterable of int
        A collection of integers.

    sort_key : callable, optional
        Callable used to sort the results 
        in the case of multiple perfect squares.

        By default, the function returns the lowest perfect square.

    reverse_on_key : bool, optional
        Whether to reverse the sort order
        in the case of multiple squares.
    
        When ``False`` is used with the default ``sort_key``,
        the highest perfect square will be returned.

    Examples
    --------

    >>> squarest_of([1,2,3,4,5,6,7,8,9,10,11,12])
    1

    >>> squarest_of([1,2,3,4,5,6,7,8,9,10,11,12],reverse_on_key=True)
    9

    >>> squarest_of([8,10,15,17,24,26,35,37,48,50,63,65])
    63

    >>> squarest_of([x for x in range(2,8000) if is_prime(x)])
    2
    """

    try:
        # Get a perfect square.
        return sorted([x for x in c if is_square(x)], key=sort_key, reverse=reverse_on_key)[0]
    except IndexError:
        # There weren't any perfect squares. Get the squarest.
        return max(c, key=lambda x:squareness(x))
            
        
@functools.lru_cache()
def squarish_factors(x):
    """Returns the closest pair of factors of x.

    Parameters
    ----------

    x : int

    Examples
    --------

    >>> squarish_factors(20)
    (5, 4)

    >>> squarish_factors(36)
    (6, 6)

    >>> squarish_factors(53)
    (53, 1)

    >>> squarish_factors(0)
    (0, 0)

    """
    if is_square(x):
        sr = int(math.sqrt(x))
        return sr, sr
    
    mid = math.floor(math.sqrt(x))

    for d in range(mid,0,-1):
        if x%d == 0:
            return int(x/d),d

    # if the above loop completes, there's a problem
    raise ValueError("X must be a positive integer.")



def is_squarish(x, threshold=0.75):
    """Determines whether x is sufficiently squarish.

    Parameters
    ----------

    x : int

    threshold : float, optional
        The lower bound of acceptable squareness.

    Examples
    --------

    >>> is_squarish(67)
    False

    >>> is_squarish(12)
    True
    """

    return squareness(x) > threshold


def squareness(x):
    """Returns a number between ``(0,1)`` 
    representing the squareness of ``x``.

    All perfect squares have a squareness of ``1``.

    Examples
    --------

    >>> squareness(1)
    1.0

    >>> squareness(4)
    1.0

    >>> squareness(5) < squareness(12)
    True
    

    """

    
    if x == 0:
        return 1
    sr = math.sqrt(x)
    m,n = squarish_factors(x)
    
    return (sr/m + n/sr)/2

    
    



def r_size(x):
    """Returns the difference between the largest and smallest value in a collection.

    Parameters
    ----------

    x : collection of numbers
    
    Examples
    --------

    >>> r_size([2,4,6,8])
    6

    >>> r_size({3,6,9,11})
    8

    """
    return max(x) - min(x)


def is_square(x):
    """Returns True if x is a perfect square. Otherwise False.

    Examples
    --------
    
    >>> is_square(1)
    True

    >>> is_square(-1)
    False

    >>> is_square(4)
    True
    """

    try:
        return int(math.sqrt(x)) == math.sqrt(x)
    except ValueError:
        return False


def is_prime(x):
    """Returns True if x is prime. Otherwise False.

    Examples
    --------

    >>> is_prime(1)
    True

    >>> is_prime(2)
    True

    >>> is_prime(3)
    True

    >>> is_prime(4)
    False

    >>> is_prime(5)
    True

    >>> is_prime(6)
    False

    >>> is_prime(7)
    True

    """

    return min(squarish_factors(x)) == 1

